From d36dc0bcf5a09059cb63f96fb5765a88bf1135e8 Mon Sep 17 00:00:00 2001
From: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Date: Wed, 26 Sep 2018 16:35:23 +0300
Subject: [PATCH 4738/5242] MLK-19743 sound: asoc: fix micfill sleep in
 interrupt context

commit  828607601bc5011d6f0fb989176f013c93bad11b from
https://source.codeaurora.org/external/imx/linux-imx.git

There is a problem when an interrupt is handled by hwvad_isr
since it calls kobject_uevent_env which can sleep.
Solved it by doing the kobject_uevent_env in a threaded_irq
function.

Signed-off-by: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Reviewed-by: Shengjiu Wang <shengjiu.wang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_micfil.c |   14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/sound/soc/fsl/fsl_micfil.c b/sound/soc/fsl/fsl_micfil.c
index 61cbb99..e04dd20 100644
--- a/sound/soc/fsl/fsl_micfil.c
+++ b/sound/soc/fsl/fsl_micfil.c
@@ -1763,8 +1763,9 @@ static bool fsl_micfil_volatile_reg(struct device *dev, unsigned int reg)
 
 /* END OF REGMAP */
 
-static void voice_detected_irq(struct fsl_micfil *micfil)
+static irqreturn_t voice_detected_fn(int irq, void *devid)
 {
+	struct fsl_micfil *micfil = (struct fsl_micfil *)devid;
 	struct device *dev = &micfil->pdev->dev;
 	int ret;
 
@@ -1780,6 +1781,8 @@ static void voice_detected_irq(struct fsl_micfil *micfil)
 
 	/* notify userspace that voice was detected */
 	kobject_uevent_env(&dev->kobj, KOBJ_CHANGE, envp);
+
+	return IRQ_HANDLED;
 }
 
 static irqreturn_t hwvad_isr(int irq, void *devid)
@@ -1805,10 +1808,9 @@ static irqreturn_t hwvad_isr(int irq, void *devid)
 				   MICFIL_VAD0_STAT_IF_MASK,
 				   MICFIL_VAD0_STAT_IF);
 
-		voice_detected_irq(micfil);
 	}
 
-	return IRQ_HANDLED;
+	return IRQ_WAKE_THREAD;
 }
 
 static irqreturn_t hwvad_err_isr(int irq, void *devid)
@@ -2136,9 +2138,9 @@ static int fsl_micfil_probe(struct platform_device *pdev)
 	/* Digital Microphone interface voice activity detector event
 	 * interrupt - IRQ 44
 	 */
-	ret = devm_request_irq(&pdev->dev, micfil->irq[0],
-			       hwvad_isr, irqflag,
-			       micfil->name, micfil);
+	ret = devm_request_threaded_irq(&pdev->dev, micfil->irq[0],
+					hwvad_isr, voice_detected_fn,
+					irqflag, micfil->name, micfil);
 	if (ret) {
 		dev_err(&pdev->dev, "failed to claim hwvad event irq %u\n",
 			micfil->irq[0]);
-- 
1.7.9.5

